--- gcc-9.3.0/gcc/config/mips/mips.opt	2020-03-12 11:07:21.000000000 +0000
+++ gcc-patches/mips.opt	2020-07-22 11:31:13.107604123 +0000
@@ -475,3 +475,15 @@
 mloongson-ext2
 Target Report Var(TARGET_LOONGSON_EXT2)
 Use Loongson EXTension R2 (EXT2) instructions.
+
+mno-unaligned-mem-access
+Target Report Mask(NO_UNALIGNED_MEM_ACCESS)
+Do not use patented lwl/lwr/swl/swr instructions
+
+mnohwdiv
+Target Report Mask(NO_HW_DIV)
+No hardware divu/div instruction support
+
+mnohwmult
+Target Report Mask(NO_HW_MULT)
+No hardware multiplier instruction support
